### ISSUE TITLE
[Request] Customize how certain symbols are highlighted via custom hooks

### ISSUE THREAD

**Issue by kobalicek** [highlightjs/highlight.js#3829]
**Is your request related to a specific problem you're having?**

Possibly.

**The solution you'd prefer / feature you'd like to see added...**

I would like to be able to select how certain symbols are highlighted without having to create a completely new language. 

At the moment, I'm using highlight.js to highlight sample code written in C++. In all cases, this code is demonstrating how to use a particular C++ library. I would like to teach highlight.js to highlight symbols that are exposed by such library. I have all the symbols the library exposed - functions, classes, constants, etc...

What I would like to have is some kind of a hook that can be called to override the default. For example if I see a symbol "BlaBla" I would like to tell highlight.js to use a specific class for that symbol and not the default or nothing.

**Any alternative solutions you considered...**

I have explored highlight.js plugin API, but it seems it's not suitable for this particular feature. For example a function parameter is shown as "SomeType* parameter" - which mixes multiple tokens, so I would have to do additional parsing to separate types from variable names, etc... What I need is a token based hook instead.

I'm exploring Prism.js and monaco editor at the moment - both seem to have this feature. Not decided to switch immediately, just wanted to know whether I can hook highlight.js first.

**Comment by joshgoebel** [comment]
I don't think this is something we're interested in supporting in core. 

I also truly don't think it would work well for many use cases.  The lexing and parsing rules used at any given time are defined by the language's grammar file... ie much of the content in a large body of text might be skipped over entirely... the only tokens that exist are ones explicitly recognized by a grammar. Your example:

`SomeType* parameter`

Is a good example... if that's what the grammar declares as a single scope, then that is effectively a single token - it's not 3 (or 4) tokens you could do with as you please - without doing your own further sub-parsing.

---

That said there are two things perhaps worth looking into:

- `Language#__emitTokens` - https://highlightjs.readthedocs.io/en/latest/mode-reference.html#emittokens (beta API, going public soon) [per grammar]
- `Configuration#_emitter` (private) [global]

The first allows you to essentially build your own custom parsing/tokenizing engine inside Highlight.js... so much you can do with this... but you're starting from ground zero... you parse and tokenize code yourself - then emit those tokens into the HLJS engine.

The latter private API allows you to replace the built-in `TokenTreeEmitter` with a custom emitter... it would allow you to do things like easily change the type of scope that was registered for particular blocks, etc... This is all private API (though a remarkably stable one), hence not supported here, so I won't go into further detail... but the source should be reasonably easy to follow if someone really wanted to experiment at that level.

**Comment by kobalicek** [comment]
Hello, thanks for the response!

I have migrated to prism.js and I was able to do what I wanted without having to define my own language. I have just used prism.js hook interface to influence the colorization the way I needed.

I'm attaching a screenshot of what I wanted to achieve:

![image](https://github.com/highlightjs/highlight.js/assets/6491301/5f28ea18-38e0-4ca9-92e0-1ae96ae7bc39)

I think this can be closed if it's not something that is planned.
